cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/plugin/PluginInstantiator.java

```diff
@@ -211,22 +220,57 @@ public ClassLoader getArtifactClassLoader(ArtifactId artifactId) throws IOExcept
     }
   }
 
-  private PluginProperties substituteMacroFields(Plugin plugin, MacroEvaluator macroEvaluator) {
+  private PluginProperties substituteMacros(Plugin plugin, @Nullable MacroEvaluator macroEvaluator) {
     Map<String, String> properties = new HashMap<>();
     Map<String, PluginPropertyField> pluginPropertyFieldMap = plugin.getPluginClass().getProperties();
-    MacroParser macroParser = new MacroParser(macroEvaluator);
+
+    // create macro evaluator and parser based on if it is config or runtime
+    boolean configTime = macroEvaluator == null;
+    TrackingMacroEvaluator trackingMacroEvaluator = new TrackingMacroEvaluator();
+    MacroParser macroParser;
+    if (configTime) {
+      macroParser = new MacroParser(trackingMacroEvaluator);
+    } else {
+      macroParser = new MacroParser(macroEvaluator);
+    }
 
     for (Map.Entry<String, String> property : plugin.getProperties().getProperties().entrySet()) {
       PluginPropertyField field = pluginPropertyFieldMap.get(property.getKey());
+      String propertyValue = property.getValue();
       if (field != null && field.isMacroSupported()) {
-        properties.put(property.getKey(), macroParser.parse(property.getValue()));
-      } else {
-        properties.put(property.getKey(), property.getValue());
+        if (configTime) {
+          // parse for syntax check and check if trackingMacroEvaluator finds macro syntax present
+          macroParser.parse(propertyValue);
+          propertyValue = getOriginalOrDefaultValue(propertyValue, property.getKey(), field.getType(),
+                                                    trackingMacroEvaluator);
+        } else {
+          propertyValue = macroParser.parse(propertyValue);
+        }
       }
+      properties.put(property.getKey(), propertyValue);
     }
     return PluginProperties.builder().addAll(properties).build();
   }
 
+  private String getOriginalOrDefaultValue(String originalPropertyString, String propertyName, String propertyType,
+                                           TrackingMacroEvaluator trackingMacroEvaluator) {
+    if (trackingMacroEvaluator.hasMacro()) {
+      trackingMacroEvaluator.reset();
+      return getDefaultProperty(propertyName, propertyType);
+    }
+    return originalPropertyString;
+  }
+
+  private String getDefaultProperty(String propertyName, String propertyType) {
+    Class propertyClass = PROPERTY_TYPES.get(propertyType);
+    if (propertyClass == null) {
+      throw new IllegalArgumentException(String.format("Unable to get default value for property %s of type %s.",
+                                                       propertyName, propertyType));
+    }
+    Object defaultProperty = Defaults.defaultValue(propertyClass);
+    return defaultProperty == null ? null : defaultProperty.toString();
+  }
+
   private Set<String> getFieldsWithMacro(Plugin plugin) {
```

Comment:
this method can also be cleaned up quite a bit after https://github.com/caskdata/cdap/pull/6089 is merged. Please also add a comment with a todo and make sure @gsps1 knows about it.